app/services/concerns/xccdf/rule_group_rules.rb

```diff
@@ -0,0 +1,119 @@
+# frozen_string_literal: true
+
+module Xccdf
+  # Methods related to saving RuleGroupRules
+  module RuleGroupRules
+    extend ActiveSupport::Concern
+
+    included do
+      def save_rule_group_rules
+        @op_rules_and_rule_groups = @op_rule_groups + @op_rules
+        @rule_group_rules = all_rule_group_rules
+
+        ::RuleGroupRule.import!(
+          @rule_group_rules.select(&:new_record?), ignore: true
+        )
+
+        rgr_base = ::RuleGroupRule.left_outer_joins(dependent_rule_group: :benchmark)
+                                  .where.not(conflicting_id: nil)
+                                  .or(::RuleGroupRule.left_outer_joins(dependent_rule_group: :benchmark)
+                                  .where.not(requiring_id: nil))
+        rgr_links_to_remove(rgr_base).delete_all
+      end
+
+      def save_rule_parents
+        @rule_group_rules = rules_with_rule_group_parent
+        ::RuleGroupRule.import!(@rule_group_rules.select(&:new_record?),
+                                on_duplicate_key_update: {
+                                  conflict_target: %i[rule_id],
+                                  columns: %i[rule_group_id]
+                                })
+
+        rule_parents_base = ::RuleGroupRule.joins(rule: :benchmark)
+
+        rule_parent_links_to_remove(rule_parents_base).delete_all
+      end
+
+      def save_all_rule_group_rules
+        save_rule_group_rules
+        save_rule_parents
+      end
+
+      private
+
+      def all_rule_group_rules
+        @all_rule_group_rules ||= @op_rules_and_rule_groups.flat_map do |op_r_or_rg|
+          with_conflicts(op_r_or_rg) + with_requires(op_r_or_rg)
+        end.compact
+      end
+
+      def with_conflicts(op_r_or_rg)
+        depending = rule_or_rule_group_for(ref_id: op_r_or_rg.id)
+        op_r_or_rg.conflicts.map do |conflicting_ref_id|
+          conflicting = rule_or_rule_group_for(ref_id: conflicting_ref_id)
+          next unless conflicting
+
+          ::RuleGroupRule.find_or_initialize_by(
+            depending: depending, conflicting: conflicting
+          )
+        end
+      end
+
+      def with_requires(op_r_or_rg)
+        depending = rule_or_rule_group_for(ref_id: op_r_or_rg.id)
+        op_r_or_rg.requires.map do |required_ref_id|
+          requiring = rule_or_rule_group_for(ref_id: required_ref_id)
+          next unless requiring
+
+          ::RuleGroupRule.find_or_initialize_by(
+            depending: depending, requiring: requiring
+          )
+        end
+      end
+
+      def rules_with_rule_group_parent
+        @rules_with_rule_group_parent ||= @op_rules.flat_map do |op_rule|
+          rule_group = rule_group_for(ref_id: op_rule.parent_id)
+          next unless rule_group
+
+          rule = rule_for(ref_id: op_rule.id)
+          ::RuleGroupRule.find_or_initialize_by(
```

Comment:
Maybe you can reuse the cache here similarly to as I did in the saving of the parents of `RuleGroup` below.

Reply:
Yeah, I'll try to do something like that here too